pennylane/transforms/optimization/commute_through_controls_targets.py

```diff
@@ -0,0 +1,155 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Transforms for pushing commuting gates through targets/control qubits."""
+
+from itertools import chain
+
+from pennylane import apply
+from pennylane.wires import Wires
+from pennylane.transforms import qfunc_transform
+
+from .optimization_utils import find_next_gate
+
+# Gates that commute among each particular basis
+commuting_gates = {
+    "X": ["PauliX", "RX", "SX"],
+    "Y": ["PauliY", "RY"],
+    "Z": ["PauliZ", "RZ", "PhaseShift", "S", "T"],
+}
+
+movable_gates = list(chain(*commuting_gates.values()))
+
+
+@qfunc_transform
+def commute_through_controls_targets(tape):
+    """Quantum function transform to move commuting gates past
+    control and target qubits of controlled operations.
+
+    Args:
+        qfunc (function): A quantum function.
```

Comment:
I can't recall if other compilation transforms have this, but does it make sense to have

```
Returns:
    function: the transformed quantum function
```

Reply:
They don't, but I think that's a good idea. Will update!